src/lib/PageTracker.js

```diff
@@ -0,0 +1,410 @@
+    /** @property {boolean} - Whether Advanced Content Insights is on for the file that we are tracking */
+    active = false;
+
+    /** @property {number|null} - Current page number the document is at */
+    currentPage;
+
+    /** @property {HTMLElement} - Preview DOM container */
+    container;
+
+    /** @property {Array} - Events to be reported */
+    events = [];
+
+    /** @property {Object} - Current file being previewed */
+    file = {};
+
+    /** @property {number} - File Length. On Document file types will be the number of pages. */
+    fileLength;
+
+    /** @property {string} - File Owner Enterprise ID. */
+    fileOwnerEId = '';
+
+    /** @property {number} -  Idle Timeout ID */
+    idleTimerTimeoutId;
+
+    /** @property {boolean} - Whether the Access Stats Preview Event was reported */
+    previewEventReported = false;
+
+    /** @property {number} - Interval ID for reporting page data */
+    reportingIntervalId;
+
+    /** @property {string|null} - Random ID to identify a particular Session/Visit */
+    sessionId;
+
+    /** @property {Date|null} - Reference date from where the current report to be sent starts */
+    startTime;
+
+    /** @property {string} - User Enterprise ID */
+    userEId = '';
+
+    /** @property {boolean} - Whether the user events are binded or not */
+    userEventsBounded = false;
+
+    /** @property {number|null} - User ID */
+    userId;
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * [constructor]
+     * @param {Object} options - Content Insights Related Options
+     * @param {Object} file - File that is being previewed
+     * @return {PageTracker} PageTracker instance
+     */
+    constructor(options = {}, file) {
+        super();
+        this.setOptions(options);
+        this.file = file;
+        this.activeUserHandler = this.activeUserHandler.bind(this);
+        this.inactiveUserHandler = this.inactiveUserHandler.bind(this);
+        this.throttledActiveUserHandler = throttle(this.activeUserHandler, ACTIVE_USER_THROTTLE, { trailing: false });
+    }
+
+    /**
+     * [destructor]
+     *
+     * @return {void}
+     */
+    destroy() {
+        this.stopTracking();
+        this.unbindIdleUserEvents();
+        this.clearIdleTimeout();
+        this.currentPage = null;
+        this.file = {};
+        this.sessionId = null;
+        this.startTime = null;
+    }
+
+    /**
+     * Get the session id
+     *
+     * @public
+     * @return {string}
+     */
+    getSessionId() {
+        return this.sessionId;
+    }
+
+    /**
+     * Call the interval starting function and set the idle timer
+     *
+     * @public
+     * @return {void}
+     */
+    init() {
+        // Generate a new session id. This would need to be later changed to be generated on server side
+        if (!this.sessionId) {
```

Comment:
Preview is often destroyed/recreated frequently based on user actions. Does this session id persist adequately or does it need to be cached somehow?

Reply:
I think the sessionId doesn't need to be persisted. As far as I can see, each time Preview is destroyed or recreated, a new preview event request is made to Access Stats. Since a new preview event was sent, we also need to send the new sessionId. Do we have any use cases where Preview is destroyed and a new access stats request is not being made?